File: tests/performance/harness/strict_json.mjs
import { open } from "node:fs/promises";

export const JSON_LIMITS = Object.freeze({
  maxBytes: 128 * 1024 * 1024,
  maxDepth: 32,
  maxCollectionEntries: 100_000,
  maxTotalValues: 4_000_000,
  maxStringBytes: 65_536,
});

export async function readStrictJson(path, { maxBytes = JSON_LIMITS.maxBytes } = {}) {
  const handle = await open(path, "r");
  try {
    const metadata = await handle.stat();
    if (metadata.size > maxBytes) throw new SyntaxError(`JSON input exceeds ${maxBytes} bytes`);

    const chunks = [];
    let total = 0;
    while (true) {
      const chunk = Buffer.allocUnsafe(Math.min(64 * 1024, maxBytes + 1 - total));
      const { bytesRead } = await handle.read(chunk, 0, chunk.length, null);
      if (bytesRead === 0) break;
      total += bytesRead;
      if (total > maxBytes) throw new SyntaxError(`JSON input exceeds ${maxBytes} bytes`);
      chunks.push(chunk.subarray(0, bytesRead));
    }
    const text = new TextDecoder("utf-8", { fatal: true }).decode(Buffer.concat(chunks, total));
    return parseStrictJson(text, { maxBytes });
  } finally {
    await handle.close();
  }
}

export function parseStrictJson(source, { maxBytes = JSON_LIMITS.maxBytes } = {}) {
  if (typeof source !== "string") throw new SyntaxError("JSON input must be a string");
  if (Buffer.byteLength(source, "utf8") > maxBytes) throw new SyntaxError(`JSON input exceeds ${maxBytes} bytes`);
  new Scanner(source).validate();
  return JSON.parse(source);
}

class Scanner {
  constructor(source) {
    this.source = source;
    this.index = 0;
    this.values = 0;
  }

  validate() {
    this.whitespace();
    this.value(0);
    this.whitespace();
    if (!this.eof()) this.fail("trailing content");
  }

  value(depth) {
    this.values += 1;
    if (this.values > JSON_LIMITS.maxTotalValues) this.fail(`JSON value cardinality exceeds ${JSON_LIMITS.maxTotalValues}`);
    if (depth > JSON_LIMITS.maxDepth) this.fail(`JSON nesting exceeds ${JSON_LIMITS.maxDepth}`);

    switch (this.current()) {
      case "{": return this.object(depth + 1);
      case "[": return this.array(depth + 1);
      case '"': return this.string(false);
      case "t": return this.literal("true");
      case "f": return this.literal("false");
      case "n": return this.literal("null");
      default:
        if (this.current() === "-" || this.isDigit(this.current())) return this.number();
        this.fail("expected JSON value");
    }
  }

  object(depth) {
    this.consume("{");
    this.whitespace();
    if (this.current() === "}") return this.consume("}");
    const keys = new Set();
    let entries = 0;
    while (true) {
      if (this.current() !== '"') this.fail("object key must be a string");
      const key = this.string(true);
      if (keys.has(key)) this.fail("duplicate object key");
      keys.add(key);
      entries += 1;
      if (entries > JSON_LIMITS.maxCollectionEntries) this.fail(`object cardinality exceeds ${JSON_LIMITS.maxCollectionEntries}`);
      this.whitespace();
      this.consume(":");
      this.whitespace();
      this.value(depth);
      this.whitespace();
      if (this.current() === "}") return this.consume("}");
      this.consume(",");
      this.whitespace();
    }
  }

  array(depth) {
    this.consume("[");
    this.whitespace();
    if (this.current() === "]") return this.consume("]");
    let entries = 0;
    while (true) {
      entries += 1;
      if (entries > JSON_LIMITS.maxCollectionEntries) this.fail(`array cardinality exceeds ${JSON_LIMITS.maxCollectionEntries}`);
      this.value(depth);
      this.whitespace();
      if (this.current() === "]") return this.consume("]");
      this.consume(",");
      this.whitespace();
    }
  }

  string(capture) {
    this.consume('"');
    let decoded = "";
    let decodedBytes = 0;
    while (true) {
      if (this.eof()) this.fail("unterminated string");
      const code = this.source.charCodeAt(this.index);
      if (code === 0x22) {
        this.index += 1;
        return capture ? decoded : undefined;
      }
      if (code < 0x20) this.fail("unescaped control character");

      let segment;
      if (code === 0x5c) {
        this.index += 1;
        const escaped = this.current();
        const simple = { '"': '"', "\\": "\\", "/": "/", b: "\b", f: "\f", n: "\n", r: "\r", t: "\t" };
        if (Object.hasOwn(simple, escaped)) {
          segment = simple[escaped];
          this.index += 1;
        } else if (escaped === "u") {
          segment = this.unicodeEscape();
        } else {
          this.fail("invalid string escape");
        }
      } else if (code < 0x80) {
        segment = capture ? this.source[this.index] : "";
        this.index += 1;
        decodedBytes += 1;
        if (decodedBytes > JSON_LIMITS.maxStringBytes) this.fail(`decoded string exceeds ${JSON_LIMITS.maxStringBytes} bytes`);
        if (capture) decoded += segment;
        continue;
      } else if (code >= 0xd800 && code <= 0xdbff) {
        const low = this.source.charCodeAt(this.index + 1);
        if (!(low >= 0xdc00 && low <= 0xdfff)) this.fail("unpaired Unicode surrogate");
        segment = this.source.slice(this.index, this.index + 2);
        this.index += 2;
      } else {
        if (code >= 0xdc00 && code <= 0xdfff) this.fail("unpaired Unicode surrogate");
        segment = this.source[this.index];
        this.index += 1;
      }
      decodedBytes += Buffer.byteLength(segment, "utf8");
      if (decodedBytes > JSON_LIMITS.maxStringBytes) this.fail(`decoded string exceeds ${JSON_LIMITS.maxStringBytes} bytes`);
      if (capture) decoded += segment;
    }
  }

  unicodeEscape() {
    this.consume("u");
    const first = this.hexCodeUnit();
    let codepoint = first;
    if (first >= 0xd800 && first <= 0xdbff) {
      this.consume("\\");
      this.consume("u");
      const second = this.hexCodeUnit();
      if (!(second >= 0xdc00 && second <= 0xdfff)) this.fail("invalid Unicode surrogate pair");
      codepoint = 0x10000 + ((first - 0xd800) << 10) + (second - 0xdc00);
    } else if (first >= 0xdc00 && first <= 0xdfff) {
      this.fail("unpaired Unicode surrogate");
    }
    return String.fromCodePoint(codepoint);
  }

  hexCodeUnit() {
    const digits = this.source.slice(this.index, this.index + 4);
    if (!/^[0-9A-Fa-f]{4}$/.test(digits)) this.fail("invalid Unicode escape");
    this.index += 4;
    return Number.parseInt(digits, 16);
  }

  number() {
    if (this.current() === "-") this.index += 1;
    if (this.current() === "0") {
      this.index += 1;
      if (this.isDigit(this.current())) this.fail("leading zero in number");
    } else {
      if (!/[1-9]/.test(this.current() ?? "")) this.fail("invalid number");
      while (this.isDigit(this.current())) this.index += 1;
    }
    if (this.current() === ".") {
      this.index += 1;
      if (!this.isDigit(this.current())) this.fail("invalid fractional number");
      while (this.isDigit(this.current())) this.index += 1;
    }
    if (this.current() === "e" || this.current() === "E") {
      this.index += 1;
      if (this.current() === "+" || this.current() === "-") this.index += 1;
      if (!this.isDigit(this.current())) this.fail("invalid exponent");
      while (this.isDigit(this.current())) this.index += 1;
    }
  }

  literal(expected) {
    if (this.source.slice(this.index, this.index + expected.length) !== expected) this.fail("invalid literal");
    this.index += expected.length;
  }

  whitespace() {
    while ([" ", "\t", "\n", "\r"].includes(this.current())) this.index += 1;
  }

  consume(expected) {
    if (this.current() !== expected) this.fail(`expected ${JSON.stringify(expected)}`);
    this.index += 1;
  }

  current() { return this.source[this.index]; }
  eof() { return this.index >= this.source.length; }
  isDigit(value) { return value !== undefined && value >= "0" && value <= "9"; }
  fail(message) { throw new SyntaxError(`${message} at character ${this.index}`); }
}

File: tests/performance/harness/strict_json.rb
# frozen_string_literal: true

require "json"
require "pathname"
require "set"

module Stead
  module PerformanceStrictJSON
    MAX_JSON_BYTES = 128 * 1024 * 1024
    MAX_DEPTH = 32
    MAX_COLLECTION_ENTRIES = 100_000
    MAX_TOTAL_VALUES = 4_000_000
    MAX_STRING_BYTES = 65_536

    class Error < JSON::ParserError; end

    module_function

    def parse_file(path, max_bytes: MAX_JSON_BYTES)
      path = Pathname(path)
      size = path.size
      raise Error, "JSON input exceeds #{max_bytes} bytes" if size > max_bytes

      bytes = File.open(path, "rb") { |file| file.read(max_bytes + 1) }
      raise Error, "JSON input exceeds #{max_bytes} bytes" if bytes.bytesize > max_bytes

      parse(bytes)
    end

    def parse(source, max_bytes: MAX_JSON_BYTES)
      raise Error, "JSON input must be a string" unless source.is_a?(String)
      raise Error, "JSON input exceeds #{max_bytes} bytes" if source.bytesize > max_bytes

      source = source.dup.force_encoding(Encoding::UTF_8)
      raise Error, "JSON input is not valid UTF-8" unless source.valid_encoding?

      Scanner.new(source).validate!
      JSON.parse(source, max_nesting: MAX_DEPTH)
    rescue JSON::NestingError => error
      raise Error, error.message
    end

    class Scanner
      SIMPLE_ESCAPES = {
        0x22 => '"', 0x5c => "\\", 0x2f => "/", 0x62 => "\b",
        0x66 => "\f", 0x6e => "\n", 0x72 => "\r", 0x74 => "\t"
      }.freeze

      def initialize(source)
        @source = source
        @index = 0
        @values = 0
      end

      def validate!
        whitespace
        value(0)
        whitespace
        fail_at("trailing content") unless eof?
        true
      end

      private

      def value(depth)
        @values += 1
        fail_at("JSON value cardinality exceeds #{MAX_TOTAL_VALUES}") if @values > MAX_TOTAL_VALUES
        fail_at("JSON nesting exceeds #{MAX_DEPTH}") if depth > MAX_DEPTH

        case byte
        when 0x7b then object(depth + 1)
        when 0x5b then array(depth + 1)
        when 0x22 then string(capture: false)
        when 0x74 then literal("true")
        when 0x66 then literal("false")
        when 0x6e then literal("null")
        when 0x2d, 0x30..0x39 then number
        else fail_at("expected JSON value")
        end
      end

      def object(depth)
        consume(0x7b)
        whitespace
        return consume(0x7d) if byte == 0x7d

        keys = Set.new
        entries = 0
        loop do
          fail_at("object key must be a string") unless byte == 0x22
          key = string(capture: true)
          fail_at("duplicate object key") unless keys.add?(key)
          entries += 1
          fail_at("object cardinality exceeds #{MAX_COLLECTION_ENTRIES}") if entries > MAX_COLLECTION_ENTRIES
          whitespace
          consume(0x3a)
          whitespace
          value(depth)
          whitespace
          break consume(0x7d) if byte == 0x7d

          consume(0x2c)
          whitespace
        end
      end

      def array(depth)
        consume(0x5b)
        whitespace
        return consume(0x5d) if byte == 0x5d

        entries = 0
        loop do
          entries += 1
          fail_at("array cardinality exceeds #{MAX_COLLECTION_ENTRIES}") if entries > MAX_COLLECTION_ENTRIES
          value(depth)
          whitespace
          break consume(0x5d) if byte == 0x5d

          consume(0x2c)
          whitespace
        end
      end

      def string(capture:)
        consume(0x22)
        decoded = +"" if capture
        decoded_bytes = 0
        loop do
          fail_at("unterminated string") if eof?
          current = byte
          if current == 0x22
            @index += 1
            return decoded
          end
          fail_at("unescaped control character") if current < 0x20

          if current == 0x5c
            @index += 1
            escaped = byte
            if SIMPLE_ESCAPES.key?(escaped)
              decoded << SIMPLE_ESCAPES.fetch(escaped) if capture
              decoded_bytes += 1
              @index += 1
            elsif escaped == 0x75
              character = unicode_escape
              decoded << character if capture
              decoded_bytes += character.bytesize
            else
              fail_at("invalid string escape")
            end
          elsif current < 0x80
            decoded << current if capture
            decoded_bytes += 1
            @index += 1
          else
            width = utf8_width(current)
            fail_at("invalid UTF-8 in string") if width.zero? || @index + width > @source.bytesize
            segment = @source.byteslice(@index, width)
            fail_at("invalid UTF-8 in string") unless segment.dup.force_encoding(Encoding::UTF_8).valid_encoding?
            decoded << segment if capture
            decoded_bytes += width
            @index += width
          end
          fail_at("decoded string exceeds #{MAX_STRING_BYTES} bytes") if decoded_bytes > MAX_STRING_BYTES
        end
      end

      def unicode_escape
        consume(0x75)
        first = hex_code_unit
        codepoint = if first.between?(0xd800, 0xdbff)
                      consume(0x5c)
                      consume(0x75)
                      second = hex_code_unit
                      fail_at("invalid Unicode surrogate pair") unless second.between?(0xdc00, 0xdfff)
                      0x10000 + ((first - 0xd800) << 10) + (second - 0xdc00)
                    else
                      fail_at("unpaired Unicode surrogate") if first.between?(0xdc00, 0xdfff)
                      first
                    end
        [codepoint].pack("U")
      end

      def hex_code_unit
        fail_at("truncated Unicode escape") if @index + 4 > @source.bytesize
        digits = @source.byteslice(@index, 4)
        fail_at("invalid Unicode escape") unless digits.match?(/\A[0-9A-Fa-f]{4}\z/)
        @index += 4
        digits.to_i(16)
      end

      def number
        @index += 1 if byte == 0x2d
        fail_at("invalid number") if eof?
        if byte == 0x30
          @index += 1
          fail_at("leading zero in number") if digit?(byte)
        else
          fail_at("invalid number") unless byte&.between?(0x31, 0x39)
          @index += 1 while digit?(byte)
        end
        if byte == 0x2e
          @index += 1
          fail_at("invalid fractional number") unless digit?(byte)
          @index += 1 while digit?(byte)
        end
        if byte == 0x65 || byte == 0x45
          @index += 1
          @index += 1 if byte == 0x2b || byte == 0x2d
          fail_at("invalid exponent") unless digit?(byte)
          @index += 1 while digit?(byte)
        end
      end

      def literal(expected)
        fail_at("invalid literal") unless @source.byteslice(@index, expected.bytesize) == expected
        @index += expected.bytesize
      end

      def whitespace
        @index += 1 while [0x20, 0x09, 0x0a, 0x0d].include?(byte)
      end

      def consume(expected)
        fail_at("expected #{expected.chr.inspect}") unless byte == expected
        @index += 1
        true
      end

      def digit?(value)
        value&.between?(0x30, 0x39)
      end

      def utf8_width(first)
        return 1 if first < 0x80
        return 2 if first.between?(0xc2, 0xdf)
        return 3 if first.between?(0xe0, 0xef)
        return 4 if first.between?(0xf0, 0xf4)

        0
      end

      def byte
        @source.getbyte(@index)
      end

      def eof?
        @index >= @source.bytesize
      end

      def fail_at(message)
        raise Error, "#{message} at byte #{@index}"
      end
    end
  end
end
